src/main/java/com/github/manolo8/darkbot/gui/zones/safety/SafetiesDisplay.java

```diff
@@ -3,77 +3,123 @@
 import com.github.manolo8.darkbot.Main;
 import com.github.manolo8.darkbot.config.SafetyInfo;
 import com.github.manolo8.darkbot.gui.MapDrawer;
+import com.github.manolo8.darkbot.gui.drawables.StaticEntitiesDrawer;
+import com.github.manolo8.darkbot.gui.drawables.ZonesDrawer;
+import eu.darkbot.api.PluginAPI;
+import eu.darkbot.api.extensions.MapGraphics;
+import eu.darkbot.api.game.other.Locatable;
+import eu.darkbot.api.game.other.Point;
+import eu.darkbot.api.managers.ConfigAPI;
+import eu.darkbot.api.managers.EntitiesAPI;
+import eu.darkbot.api.managers.HeroAPI;
+import eu.darkbot.api.managers.StarSystemAPI;
 
-import java.awt.*;
 import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
 import java.util.Comparator;
 
-class SafetiesDisplay extends MapDrawer {
+public class SafetiesDisplay extends MapDrawer {
 
-    private SafetiesEditor editor;
-    private SafetyInfo closest;
+    private final SafetiesEditor editor;
+    private SafetiesDisplayDrawer<MapDrawer> safetiesDisplay;
 
     SafetiesDisplay(SafetiesEditor editor) {
         this.editor = editor;
         addMouseListener(new MouseAdapter() {
             @Override
             public void mousePressed(MouseEvent e) {
-                updateClosest(e);
-                if (closest != null) editor.edit(closest);
+                safetiesDisplay.updateClosest(e, true);
             }
         });
         addMouseMotionListener(new MouseAdapter() {
             @Override
             public void mouseMoved(MouseEvent e) {
-                updateClosest(e);
-                repaint();
+                safetiesDisplay.updateClosest(e, false);
             }
         });
     }
 
-    private void updateClosest(MouseEvent e) {
-        double x = undoTranslateX(e.getX()), y = undoTranslateY(e.getY());
-        closest = editor.safetyInfos.stream()
-                .filter(s -> s.entity != null && !s.entity.removed)
-                .min(Comparator.comparingDouble(s -> Math.pow(s.x - x, 2) + Math.pow(s.y - y, 2)))// squared distance
-                .orElse(null);
-        repaint();
-    }
-
     @Override
-    protected void paintComponent(Graphics g) {
-        Graphics2D g2 = setupDraw(g);
-        synchronized (Main.UPDATE_LOCKER) {
-            drawZones(g2);
-            drawStaticEntities(g2);
-            drawMap(g2);
-            if (editor.safetyInfos == null) return;
-            drawCustomZones(g2);
-        }
+    public void setup(Main main) {
+        super.setup(main);
+
+        this.safetiesDisplay = new SafetiesDisplayDrawer<>(main.pluginAPI, this);
+        this.safetiesDisplay.setup(editor);
```

Comment:
This class was already intentionally a very dummy class, that only relied on what the editor told it, setup shouldn't need to create its own drawer or do anything like that, simply leave it as it was, override the draw method from parent and instead of just drawing all layers, draw only what we want (zones, static entities, map name, and custom zones) by calling the different drawables required, and that's it

i don't understand the need for a SafetiesDisplayDrawer, you can simply draw ontop yourself directly in the paintComponent method